core/corehttp/proxy.go

```diff
@@ -0,0 +1,79 @@
+package corehttp
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	core "github.com/ipfs/go-ipfs/core"
+
+	p2phttp "gx/ipfs/QmPpmS1Ps3H8tBgDJWQ3pyZ2xNstzKfDd8aWUuk3Zu7muZ/go-libp2p-http"
+	protocol "gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+// ProxyOption is an endpoint for proxying a HTTP request to another ipfs peer
+func ProxyOption() ServeOption {
+	return func(ipfsNode *core.IpfsNode, _ net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
+		mux.HandleFunc("/p2p/", func(w http.ResponseWriter, request *http.Request) {
+			// parse request
+			parsedRequest, err := parseRequest(request)
+			if err != nil {
+				handleError(w, "Failed to parse request", err, 400)
+				return
+			}
+
+			request.Host = "" // Let URL's Host take precedence.
+			request.URL.Path = parsedRequest.httpPath
+			target, err := url.Parse(fmt.Sprintf("libp2p://%s", parsedRequest.target))
+			if err != nil {
+				handleError(w, "Failed to parse url", err, 400)
+				return
+			}
+
+			rt := p2phttp.NewTransport(ipfsNode.PeerHost, p2phttp.ProtocolOption(parsedRequest.name))
+			proxy := httputil.NewSingleHostReverseProxy(target)
+			proxy.Transport = rt
+			proxy.ServeHTTP(w, request)
+		})
+		return mux, nil
+	}
+}
+
+type proxyRequest struct {
+	target   string
+	name     protocol.ID
+	httpPath string // path to send to the proxy-host
+}
+
+// from the url path parse the peer-ID, name and http path
+// /p2p/$peer_id/http/$http_path
+// or
+// /p2p/$peer_id/x/$protocol/http/$http_path
+func parseRequest(request *http.Request) (*proxyRequest, error) {
```

Comment:
This function looks a bit messy and will get messier adding `/ws`.  I think now it allows things like `/x//http` (weird).


I think the cleanest is to do regexps:

```go
// /p2p/$peer_id/x/$protocol/http/$http_path
// Change to ^/p2p/(\w+)/(x/\w+/(?:http|ws))/(.*)$ when adding /ws
var longProtocolRex = regexp.MustCompile(`^/p2p/(\w+)/(x/\w+/http)/(.*)$`)

// /p2p/$peer_id/http/$http_path
// Change to `^/p2p/(\w+)/(http|ws)/(.*)$` when adding /ws
var shortProtocolRex = regexp.MustCompile(`^/p2p/(\w+)/(http)/(.*)$`)

// from the url path parse the peer-ID, name and http path
// /p2p/$peer_id/http/$http_path
// or
// /p2p/$peer_id/x/$protocol/http/$http_path
func parseRequest(request *http.Request) (*proxyRequest, error) {
	path := request.URL.Path

	matches := longProtocolRex.FindAllStringSubmatch(path, -1)
	if matches == nil {
		matches = shortProtocolRex.FindAllStringSubmatch(path, -1)
	}

	if matches == nil || len(matches) < 1 || len(matches[0]) < 4 {
		return nil, fmt.Errorf("invalid proxy request path: '%s'", path)
	}

	return &proxyRequest{
		target:   matches[0][1],
		name:     protocol.ID("/" + matches[0][2]),
		httpPath: matches[0][3],
	}, nil
}

```

Technically this would allow not having to change the function when we add `/ws`. It imposes that custom protocols have a-zA-Z0-9 characters though.

Reply:
I personally don't think regexes are cleaner. I prefer simpler, faster code. 

I also have the fast path for /http (by checking it first) which I imagine will be more common, being the default. 

Regarding ws, my understanding (I'm not an expert in them by any measure) is that a web socket starts out as a normal http request, but with a special header, which signals the server to treat it differently. If that is true, I don't understand the need for a separate /ws endpoint. Certainly for my usecase we don't need it. 

If we want to ban empty $custom protocol components then that is trivial to add.

Reply:
> I personally don't think regexes are cleaner. I prefer simpler, faster code.

I tend to agree and I think we can refactor this later if we need to.

> If that is true, I don't understand the need for a separate /ws endpoint. Certainly for my usecase we don't need it.

The `ws` case is a slightly different use-case: the websocket would be handled by the proxy and the inner stream itself would be forewarded. Thinking through that case, let's just ignore the ws stuff for now. My proposal wasn't quite right for that use-case but I think we can make it work (although it might not be completely symmetric to http).